osu.Game/Beatmaps/BeatmapManager.cs

```diff
@@ -178,6 +188,71 @@ public void Import(BeatmapSetInfo beatmapSetInfo)
         }
 
         /// <summary>
+        /// Downloads a beatmap.
+        /// </summary>
+        /// <param name="beatmapSetInfo">The <see cref="BeatmapSetInfo"/> to be downloaded.</param>
+        /// <returns>The new <see cref="DownloadBeatmapSetRequest"/>, or null if a download already exists for the same beatmap.</returns>
+        public DownloadBeatmapSetRequest Download(BeatmapSetInfo beatmapSetInfo)
+        {
+            if (api == null || downloadsMap.ContainsKey(beatmapSetInfo.OnlineBeatmapSetID)) return null;
+
+            ProgressNotification downloadNotification = new ProgressNotification
+            {
+                Text = $"Downloading {beatmapSetInfo.Metadata.Artist} - {beatmapSetInfo.Metadata.Title}",
+            };
+
+            var request = new DownloadBeatmapSetRequest(beatmapSetInfo);
+
+            request.DownloadProgressed += progress =>
+            {
+                downloadNotification.State = ProgressNotificationState.Active;
+                downloadNotification.Progress = progress;
+            };
+
+            request.Success += data =>
+            {
+                downloadNotification.State = ProgressNotificationState.Completed;
+
+                using (var stream = new MemoryStream(data))
+                using (var archive = new OszArchiveReader(stream))
+                    Import(archive);
+
+                downloadsMap.Remove(beatmapSetInfo.OnlineBeatmapSetID);
+            };
+
+            request.Failure += data =>
+            {
+                downloadNotification.State = ProgressNotificationState.Completed;
+                Logger.Error(data, "Failed to get beatmap download information");
+                downloadsMap.Remove(beatmapSetInfo.OnlineBeatmapSetID);
+            };
+
+            downloadNotification.CancelRequested += () =>
+            {
+                Logger.Log("Cancel requested");
+                downloadsMap[beatmapSetInfo.OnlineBeatmapSetID].Cancel();
+                downloadsMap.Remove(beatmapSetInfo.OnlineBeatmapSetID);
+                downloadNotification.State = ProgressNotificationState.Cancelled;
+                return true;
+            };
+
+            downloadsMap[beatmapSetInfo.OnlineBeatmapSetID] = request;
+            PostNotification?.Invoke(downloadNotification);
+
+            // don't run in the main api queue as this is a long-running task.
+            Task.Run(() => request.Perform(api));

```

Comment:
todo: ensure the `Success`/`Failure` callbacks are being scheduled to the main thread for thread safety.